### DISCUSSION TITLE
Possible Function enhancements/corrections

### DISCUSSION THREAD

**Discussion by bijington** [bijington/expressive#86] [Ideas]
To discuss how the existing in-built functions could be improved/or simply corrected 😄.

**Comment by graniero** [comment]
I have some notes for a few functions and support code. I'll separate them into different comments, so discussion threads are easier to track.

**Reply by bijington** [reply]
Wow thanks for all the suggestions here, they are great! It may take me some time to respond to all of them but I do have some initial thoughts that I will try and flesh out to make sure I give some accurate responses.

**Comment by graniero** [comment]
# The caret `^` as Exponent rather than Bitwise XOR
A big one is the use of the caret `^` operator for exponents rather than for bitwise XOR. Sounds like this is a pretty common topic of discussion!

For people doing a lot of bit work, it's crucial to have `^` for the standard bitwise XOR, and a lot of code could break for them if this were changed. Not a good thing!

I'm going to go out on a limb and assume that the vast majority of people who would like to use `^` for exponents are not doing bitwise operations. For those that are doing both, they are likely already using the `Pow()` function, so nothing needs to change.

I think this can be solved in a simple, non-code-breaking way via the `Context` and `ExpressiveOptions`, especially now that as of v2.2.0 you can supply a `Context` when you create a new `Expression`. I can think of two different configuration approaches, with different plusses and minuses: one may be more attractive than the other.

## Approach 1
This way replaces bitwise XOR with exponent, but leaves the other bitwise operators intact. It introduces the least change to the available operators, but it does leave the set of bitwise operators incomplete, and potentially introduces some ambiguity or uncertainty.

1. Add `CaretIsExponent = 64` to the `ExpressiveOptions` enum. (The suggested value would mesh with the suggested solution to #77.)
1. In the `Context` constructor:
``` csharp
if (Options.CaretIsExponent) {
    this.RegisterOperator(new ExponentOperator());
}
else {
    this.RegisterOperator(new BitwiseExclusiveOrOperator());
}
```

## Approach 2
This one has a larger effect on the set of available operators, but it eliminates possible ambiguity. 

1. Add `DisableBitwiseOperators = 64` to the `ExpressiveOptions` enum.
1. In the `Context` constructor:
``` csharp
if (Options.DisableBitwiseOperators) {
    this.RegisterOperator(new ExponentOperator());
}
else {
   this.RegisterOperator.(new BitwiseAndOperator());
   this.RegisterOperator.(new BitwiseOrOperator());
   this.RegisterOperator.(new BitwiseExclusiveOrOperator());
   this.RegisterOperator.(new LeftShiftOperator());
   this.RegisterOperator.(new RightShiftOperator());
}
```

**Applicable to both approaches:**
I have `ExponentOperator` and `ExponentExpression` code written that I'm happy to share.

Existing code won't break, because by default `^` remains bitwise XOR.

For `Expressions` that supply a `Context` that is constructed with `ExpressiveOptions.CaretIsExponent` (or alternatively `ExpressiveOptions.DisableBitwiseOperators` ), `^` will be the exponent operator.

For people who have some expressions doing bitwise operations and other expressions doing exponent operations, they can create two different contexts and choose the appropriate context.

For people who do bitwise operations and exponent operations, either in the same expressions or just don't want to bother with two contexts, they can stick with the default `Context` and keep using `^` for bitwise XOR  and `Pow()` for exponents.

**Reply by bijington** [reply]
I would be keen to integrate the `ExponentOperator` And `ExponentExpression`. One alternative option that I have been considering if to supply some default `Context` implementations. Probably a `BitwiseContext` and a `MathematicalContext` for now. 

Do you think that usage would be easy to follow?

**Reply by graniero** [reply]
Having a couple of default `Contexts` available is a good idea, but I’m not
sure what would exactly go into them. The `BitwiseContext` would obviously
add in the bitwise operators (at the expense of the exponent operator), but
other than having the exponent operator available, what do you see being
‘different’ in the `MathematicalContext`?

Or is that enough of a difference, and nothing else is needed? It’s
basically an alternate way of dealing with the caret, instead of using an
`ExpressionOption`. Do you see any other differences evolving between the two?

Trying to think through other impacts of this approach.... be prepared for rambling ahead...
- which one is the ‘default’ default `Context`? That is, if you just write `new
Context()`, which one do you get?
- `BitwiseContext` and `MathematicalContext` will need to be setup as
subclasses of `Context`, otherwise every time you add new ‘standard’
functions etc to the library you’ll have to remember to add them to both
contexts.
- it sets up the pattern for future ‘special’ default `Contexts`, if some new
group of functionality gets added which requires some other replacement of
operator behaviour. I can’t think of a possible one right now, but it
doesn’t mean it can’t happen!
- it gives room for adding other ‘packaged’ functions, like perhaps
`Dec2Bin()` and `Bin2Dec()` conversion functions which would maybe make sense
in a `BitwiseContext`.
  - but that just gets me thinking: the other natural functions to go with
those are `Dec2Hex()` and `Hex2Dec()`. Those would be sensible in both a
`BitwiseContext` and a `MathematicalContext`.
  - following that thought, what if you want to do the conversions because
you’re doing hex and binary representations for string display and you’re
going to do other string functions like `PadLeft()`? Which `Context` do you use
for those?
- or what if you do exponentiation AND bitwise work, but you don’t have any
need for the less commonly used XOR operator? Then you’re back to
subclassing the `BitwiseContext` to override the caret operator... which was
the original problem to solve, but with the additional complication of choosing which
default `Context` you need to use first.
- which leads me to realize that there’s extra documentation burden making
clear what’s in each default context and what’s not;
- plus library users will need to get familiar with how `Contexts` work and
how to use them early, while details of setting an `ExpressionOption` is less
documentation (initial and maintenance) and less learning.

 So partitioning into some default `Contexts` might be tricky.... maybe
`Contexts` will work better for custom, application-specific mods rather than
for default function/operator ‘bundling’.

Might be better to go with the caret operator as an `ExpressionOption` at
this point, and save choice of default `Contexts` until a stronger and more
defined set of use cases appear for partitioning?

**Reply by bijington** [reply]
It looks like replying via email possibly doesn't apply the comment to the correct discussion. Perhaps a bug to raise with GitHub...

You have summarised my thinking since writing my initials thoughts really well.

I really can’t decide if I prefer the option of supplying Contexts over options. Although I do still really like the idea of having the simple option of supplying an option.

I guess I was initially thinking that these Context defaults could work in conjunction with #1 which would allow for enabling/disabling functions by ‘type’. I wondered if it would be possible to restrict some Contexts to very small amounts of functions and then you could supply multiple or join them together in order to choose the functionality required. I think this all sounds far too overcomplicated now.

perhaps an option is going to be the simplest option here (no pun intended 😄)

**Comment by graniero** [comment]
# Validating parameters for custom functions with 0 arguments
Following discussion in #37, I have created functions for `PI()` and `E()` (which I'm happy to share).  The functions simply return the respective constant. 

When I built the custom functions I followed the built-in functions as a coding standard, but I discovered that the logic for `FunctionBase.ValidateParameterCount` did not handle the case for `expectedCount == 0` properly - it incorrectly threw an exception.

The following modification to `FunctionBase.ValidateParameterCount()` lets it verify an `expectedCount` of 0:
``` csharp
            if (expectedCount == 0 && (parameters.Any() || parameters.Length != expectedCount))
            {
                throw new ParameterCountMismatchException($"{this.Name}() does not take any arguments");
            }
            if (expectedCount != -1 && expectedCount != 0 && (parameters is null || !parameters.Any() || parameters.Length != expectedCount))
            {
                throw new ParameterCountMismatchException($"{this.Name}() takes only {expectedCount} argument(s)");
            }
```
(I'm sure these can be written more elegantly, but they get the job done. :-D)
#### My $0.02 about using functions vs variables for constants like pi and e
I think the idea of having functions for these is a better approach than introducing variables and supplying values at evaluation time:
- the expression semantics are cleaner;
- it's consistent with formula syntax in Excel, and therefore familiar to - or even assumed by! - many users if an application lets users input expressions;
- don't need to remember to supply values for the 'constant' variables at evaluation time, which is less error-prone for manual processes and less complex for coded automatic processes.

**Reply by bijington** [reply]
Oh thanks. It sounds like there is a bug with the evaluation of required parameters! I will confirm, raise the bug and use you fix. Thank you for that 😄.

As for constants I don’t know what the correct option is here. I do like the idea of the library being able to come with some built in ones. I am not sure I like them looking like a function purely because the () don’t make it read like a constant. I do agree they should have to be parsed in like a variable. Perhaps Constants should be registered much like an operator/function is and then some can be provided by default. Then the hope would be you would just have to refer to the constant simply by name.

**Reply by graniero** [reply]
A constant represented as a function, e.g. `PI()` or `E()` sure isn't a 'pure' way of representing a constant, but it's not terrible either with respect to semantics. A 'constant' function returns a value, just like any other function. The precedent for zero-argument functions is out there: for example, `random()` is a common function in many languages or applications, providing a random floating-point number between zero and one. With zero arguments, the behaviour cannot be modified but instead the behaviour is... um... constant. (sorry, couldn't resist. :laughing:) 

The other upside is very pragmatic: there is almost no work involved to introduce constants. Creating a custom function to return a fixed value is a five-minute job. Extending Expressive to handle constants as a distinct thing is a major piece of work:
- Choose a syntax for a constant: all-caps character string, or can the case be ignored? Can a constant include digits or other special characters? (probably not, that would get messy in a hurry) Does it need to be enclosed, like `Sin([Angle]*{PI}/180)`,  or can the string just be on its own, like `Sin([Angle]*PI/180)`?
  - If it's enclosed, then you need to introduce the new tokens, plus logic to make sure that what they enclose is a leaf in the tree and not another sub-expression.
  - If it's not enclosed, then you need to add peek-ahead / back-tracking to tell if you're identifying a constant or a function.
- Introduce new type of token into the parser, and do all the testing to make sure it doesn't break.
- Introduce a new Constant class, plus any other supporting classes that may end up being needed.
- Add new `RegisterConstant()` and `UnregisterConstant()` methods in `Context`, plus container variables to hold them, and all of the testing for these.
- etc.

**Reply by bijington** [reply]
Haha!

I am more than happy to agree with you here. I just tried to look up our *de-facto standard* Excel expecting `PI` to be just that but it does in fact include brackets e.g. `PI()`. The beauty of your suggestion is that in fact Expressive can remain constant (sorry, I also couldn't resist) because functions do _mostly (aside from the parameter validation you have previously mentioned)_ just work.

I guess Expressive could in fact ship with a set of these **constant functions**.

**Reply by graniero** [reply]
As I'm sure you know very well, the constant function classes are very easy to implement, but if at some point you'd like the code I wrote for `PI()` and `E()` (and the zero-parameter validation adjustment in `BaseFunction`) to save the time  just let me know your preferred workflow to receive it.

**Reply by bijington** [reply]
Cool thanks! I guess the simplest would be to provide them as a PR for #37 if you have the time.

**Reply by graniero** [reply]
No prob! I'll be back at my work machine in the next day or so (been taking a few days' break for a change!), and I'll put a PR together.

**Reply by bijington** [reply]
Awesome thanks! Enjoy the rest of your break :smile:

**Reply by bijington** [reply]
@graniero thank you for the help on this. I can confirm it is in the new release and can even be tested out here:

https://bijington.github.io/expressive-playground/

Note I do need to expand the IgnoreCase options on the playground.

**Reply by graniero** [reply]
Very happy to help, @bijington ! You have a great library here - being able to lend a hand with some light supporting tasks while you're doing the heavy lifting feels like a good way to say thank you!

Quite willing to lend a hand with other stuff that you're considering too, whenever you're ready to tackle it. I have notifications  set, so just @ mention me when you're ready for a second set of eyeballs or pair of typing hands!

**Reply by bijington** [reply]
@graniero great thank you for the offer and also the very kind words about the library 😄

**Comment by graniero** [comment]
# Behaviour of `Random()` with integer parameters
This comment just captures a behavioural quirk to which some people might respond "well yeah, of course that's what it does" while other people might respond "wow I didn't expect that", depending on what your go-to random number tool is.

When `Random(min_value, max_value)` is supplied floating-point parameters, it behaves the way you would expect based on other random-number generators: it returns a random number greater than or equal to `min_value` and less than `max_value`. The beauty of this `Random()` function is that you're not limited to [0, 1): you can give it any values and save yourself the bother of scaling and shifting [0, 1) to the range you want.

When `Random(min_value, max_value)` is supplied integer parameters, it behaves the same way: returns a random integer in the range [`min_value`, `max_value`). Nice and consistent with floating-point parameters. "well yeah, of course that's what it does".

However, if you typically do most of your random number work in Excel (yes, I know all the reasons why you wouldn't want to, but life has a way of happening) and you're used to the `RANDBETWEEN(min_value, max_value)` function, the behaviour is different than expected: unlike `RANDBETWEEN()`, you will never get your `max_value`. Easy to spot with `Random(1, 3)` but you may not notice it with `Random(1, 1000)`.  "wow I didn't expect that"

**Reply by bijington** [reply]
I am not sure I expected it :smile:.

I fully and sadly agree that Excel has created a bit of a standard way of expecting things to work, especially as most of us will probably deal with users expecting to craft expressions in the same way that Excel does. I will need to do some more investigations on what Excel offers and what exactly Expressive does. 

I do believe there should be consistency at least so perhaps this quirk will ultimately be considered a bug.

**Reply by graniero** [reply]
Yeah, it gets tricky because you've got programmers using Expressive to easily handle built-in or generated expressions within their code, which they expect to behave like the tools they're working with (e.g. their programming language); but you've also got programmers using Expressive to produce applications that easily allow users to write/design expressions within the application, which they expect to behave like the tools they're working with (e.g. Excel). In the latter case, the users writing the expressions will not be looking at Expressive's documentation - they won't even know what Expressive is!

The latter case is what I'm personally dealing with: a large operations monitoring system which allows the engineers using the system to set up a library of standard formulas/calculations to be used within or across site projects, and to set up triggering conditions for rules and alarms. For this application, and any other applications like this, users are generally going to expect the functions to work the way they're used to - and welcome to Excel as the de facto standard!

**Comment by graniero** [comment]
# Behaviour of `Count()` with string vs. number parameters
Again, this comment just captures a behavioural quirk which people should know about. However, unlike with `Random()` above, this one may warrant some discussion about what would be the most desirable behaviour.

When all of the parameters are numbers, `Count()` returns the number of parameters.

When all of the parameters are strings, `Count()` returns the total number of characters across all the strings, i.e. `Length(parameter1) + Length(parameter2) +. ....`

When the parameters are mixed, `Count()` returns the number of numeric parameters plus the total number of characters across all the string parameters:
- `Count(123) = 1`
- `Count('123') = 3`
- `Count(123, 4) = 2`
- `Count('123', 4) = 4`
- `Count(123, '4') = 2`
- `Count('foo', 123) = 4`
- `Count('foo', '123') = 6`

In all cases, the `Count()` function does not do the same thing as:
- .NET's various `Count` properties and `Count()` methods return the number of elements in the collection, or equivalently, the number of parameters in the parameter list.
- Excel's `COUNT()` function returns a count of the _numbers_ in the parameter list.

I'm not sure if the current behaviour is all that useful. I can't think of a decent use case - maybe other folks can?

The current behaviour with all string parameters is kind of like `Length()` if you were able to supply an aggregate as a parameter value like you can now do with operators. (by the way, upvote +100 for aggregates in functions, please!!!) Once that's possible, then this behaviour for string parameters would be redundant.

The current behaviour with all numeric parameters is just a count of the number of parameters, basically the same as `Count` in .NET, for example. The `Count()` function could be modified so this is the behaviour for mixed parameter types too - at least it would be consistent across types. But right now you already know how many parameters you are supplying (unless the expressions are generated, I suppose). This behaviour won't be all that powerful until you can supply an aggregate as a parameter value. (another upvote +100 for that, please!!!)

**Reply by bijington** [reply]
My initial thoughts are that this is actually a bug! 

A use case that did develop a while back was to allow for aggregate values to be counted. e.g. `Count([Prices])` where the Prices variable is in fact an array of numbers. This should give the count of numbers within the variable. I suspect somehow I have gone too far up in the inheritance tree and I am counting `IEnumerable` which `string` inherits from.

We should in theory be able to handle aggregates in *some* functions. Admittedly I may have broken that functionality and there could be other candidates that should support aggregates.

**Reply by bijington** [reply]
@graniero I have included this in a bug #99 to cover the issue because it is an issue. Good catch!

For what it is worth and again I need to apologies for a lack of documentation on this part but you can use aggregates inside some expressions like `Count`, in fact I suspect that is part of the cause of this issue.

You can call this:

```
var aggregates = new [] { 1, 2, 3, 4, 5 };
var values = new Dictionary<string, object> { ["foo"] = aggregates };
var i = new Expressive.Expression("Count([foo])");

Console.WriteLine(i.Evaluate<int>(values));
```

Also for I have found out to be quite exciting I am able to run really quick tests like this in VS Code using [dotnet interactive](https://github.com/dotnet/interactive)

![Screenshot 2021-02-02 at 21 39 35](https://user-images.githubusercontent.com/17139988/106666236-3f3b7a00-659f-11eb-8c7a-1226c7a4ca82.png)

**Reply by graniero** [reply]
@bijington I would agree with comments in #99 that `IEnumberable` is the likely culprit. Makes good sense to me.

**Comment by graniero** [comment]
I agree, I think under the current version's structure, the "caret as exponent or bitwise XOR" is most easily handled by an option - especially because it seems to be the only big operator overload issue that's come to light so far.  It's also a pretty consistent approach with the rest of the library's semantics and behaviour.

The idea of having function 'bundles' that can be added in one or many at a time is **really** cool. That adds a ton of extensibility and power in the same way that R and others do. It's a great way to deal with operator overrides, or to deal with the situation where the standard, most commonly used, or most obvious name for a function in one domain is the same name that is used in another domain to do something completely different.

However, it feels like setting up different `Contexts` to each contain different function bundles is stretching what `Context`'s responsibilities should be. When the `Context` originally started as basically the full set of functions and operators, and then someone might add or swap out one or two functions or operators, it was OK that the `Context` was not only the place to manage behavioural options but also the container for the functions, etc. But once you start getting into more involved function schemes etc., then `Context` as primary function container starts to get in the way of 'mix-and-match' or 'plug-and-play' setups.

I think another container class is needed in between the `Context` and the functions. The new class would contain a collection of functions, operators, and potentially other things (like constants?), and you would register that class or an instance of that class with your `Context` instance. Basically, the same as registering & unregistering functions or operators, but done as a cohesive group. You still end up with `Context` as the go-to spot for available functionality, but you basically start with a set of core built-in functions (managed by one of those container classes) loaded by default when the `Context` is instantiated, then build up your `Context` by loading other containers into its collection.

Either that, or if `Context` remains the container for the function and operator functionality, then you need to add a mechanism to have multiple `Contexts` active at the same time, either by chaining them together, or by adding another higher-up class to contain multiple `Contexts`. And in either approach, there is also a new need to check for function name / operator overloads/conflicts between collections, and provide a way to either select/resolve the conflict, or extend the grammar to allow a context as part of the function name....  

....starting to sound like a major version development item!

**Reply by bijington** [reply]
I agree that `Context` is probably not the correct place to be doing this if in fact it does go ahead. As you said:

> that the `Context` was not only the place to manage behavioural options but also the container for the functions

The idea to use them in the manner that I suggested would muddy the water. It could become very muddy if multiple `Context`s are supplied and dealing with things like conflicting behaviours. 

You are spot on with it sounding like some major development work! I think the option for now makes perfect sense and then if things do get extended in the manner we have discussed then the exponent behavioural differences could potentially slot in to that.

**Comment by graniero** [comment]
Yeah, I think when the time comes, `^` as exponent and `^` as bitwise XOR would very easily become part of the bundles, and any `ExpressionOptions.CaretAsExponent` value can be deprecated then obsoleted.
